smithy-model/src/main/java/software/amazon/smithy/model/shapes/ShapeId.java

```diff
@@ -59,32 +67,12 @@ private ShapeId(String namespace, String name, String member) {
     /**
      * Creates an absolute shape ID from the given string.
      *
-     * @param absoluteShapeId Shape ID to parse.
+     * @param id Shape ID to parse.
      * @return The parsed ID.
      * @throws ShapeIdSyntaxException when the ID is malformed.
      */
-    public static ShapeId from(String absoluteShapeId) {
-        int namespacePosition = absoluteShapeId.indexOf('#');
-        if (namespacePosition <= 0 || namespacePosition == absoluteShapeId.length() - 1) {
-            throw new ShapeIdSyntaxException("Invalid shape ID: " + absoluteShapeId);
-        }
-
-        String namespace = absoluteShapeId.substring(0, namespacePosition);
-        String name;
-        String memberName = null;
-
-        int memberPosition = absoluteShapeId.indexOf('$');
-        if (memberPosition == -1) {
-            name = absoluteShapeId.substring(namespacePosition + 1);
-        } else if (memberPosition < namespacePosition) {
-            throw new ShapeIdSyntaxException("Invalid shape ID: " + absoluteShapeId);
-        } else {
-            name = absoluteShapeId.substring(namespacePosition + 1, memberPosition);
-            memberName = absoluteShapeId.substring(memberPosition + 1);
-        }
-
-        validateParts(absoluteShapeId, namespace, name, memberName);
-        return new ShapeId(absoluteShapeId, namespace, name, memberName);
+    public static ShapeId from(String id) {
```

Comment:
I think the `fromParts` functions should route through this cache as well. It's used in `LoaderVisitor`  for prelude loading.

Reply:
Using the cache for that would probably be suboptimal since the check is a guess that a shape might be in the prelude, and many will not. I will add a method to shape ID called `withNamespace` to optimize this though.